### PR TITLE
resolved: fix braino with reference counting and linked lists

### LINKED ISSUE #19376: resolved: crash with > 1 search domain
**systemd version the issue has been seen with**
 > 246.13

**Used distribution**
 > Fedora 33

**Linux kernel version used** (`uname -a`)
 > `5.10.31`

**CPU architecture issue was seen on**
 > x86_64

**Expected behaviour you didn't see**
 > Able to resolve non-FQDN entries when using multiple search domains

**Unexpected behaviour you saw**
 > resolved crashes

**Steps to reproduce the problem**
 > configure > 1 search domain, but the more you configure the faster the crash

**Additional program output to the terminal or log subsystem illustrating the issue**

```text

This is similar to issue documented in pr #18407, but apparently that commit doesn't take
into account multiple search domains.  It is fairly trivial to reproduce.  The below test case
was made with 5 search domains, and a single query for a non-FQDN hostname.

Got message type=method_call sender=:1.336680 destination=org.freedesktop.resolve1 path=/org/freedesktop/resolve1 interface=org.freedesktop.resolve1.Manager member=ResolveHostname cookie=2 reply_cookie=0 signature=isit error-name=n/a error-message=n/a
idn2_lookup_u8: smbapptpgdb1 → smbapptpgdb1
Looking up RR for smbapptpgdb1 IN AAAA.
Sent message type=method_call sender=n/a destination=org.freedesktop.DBus path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=AddMatch cookie=2041 reply_cookie=0 signature=s error-name=n/a error-message=n/a
Sent message type=method_call sender=n/a destination=org.freedesktop.DBus path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=GetNameOwner cookie=2042 reply_cookie=0 signature=s error-name=n/a error-message=n/a
Got message type=method_return sender=org.freedesktop.DBus destination=:1.336171 path=n/a interface=n/a member=n/a cookie=4294967295 reply_cookie=2042 signature=s error-name=n/a error-message=n/a
Cache miss for smbapptpgdb1.searchdomain1.com IN AAAA
Transaction 55669 for <smbapptpgdb1.searchdomain1.com IN AAAA> scope dns on eth0/*.
Using feature level UDP+EDNS0 for transaction 55669.
Using DNS server 10.161.134.19 for transaction 55669.
Sending query packet with id 55669 of size 57.
Got message type=method_return sender=org.freedesktop.DBus destination=:1.336171 path=n/a interface=n/a member=n/a cookie=4294967295 reply_cookie=2041 signature= error-name=n/a error-message=n/a
Match type='signal',sender='org.freedesktop.DBus',path='/org/freedesktop/DBus',interface='org.freedesktop.DBus',member='NameOwnerChanged',arg0=':1.336680' successfully installed.
Processing incoming packet on transaction 55669 (rcode=NXDOMAIN).
Added NXDOMAIN cache entry for smbapptpgdb1.searchdomain1.com IN ANY 3600s
Transaction 55669 for <smbapptpgdb1.searchdomain1.com IN AAAA> on scope dns on eth0/* now complete with <rcode-failure> from network (unsigned).
Cache miss for smbapptpgdb1.searchdomain2.com IN AAAA
Transaction 25325 for <smbapptpgdb1.searchdomain2.com IN AAAA> scope dns on eth0/*.
Using feature level UDP+EDNS0 for transaction 25325.
Using DNS server 10.161.134.19 for transaction 25325.
Sending query packet with id 25325 of size 55.
Freeing transaction 55669.
Processing incoming packet on transaction 25325 (rcode=NXDOMAIN).
Added NXDOMAIN cache entry for smbapptpgdb1.searchdomain2.com IN ANY 3600s
Transaction 25325 for <smbapptpgdb1.searchdomain2.com IN AAAA> on scope dns on eth0/* now complete with <rcode-failure> from network (unsigned).
Cache miss for smbapptpgdb1.searchdomain3.com IN AAAA
Transaction 10284 for <smbapptpgdb1.searchdomain3.com IN AAAA> scope dns on eth0/*.
Using feature level UDP+EDNS0 for transaction 10284.
Using DNS server 10.161.134.19 for transaction 10284.
Sending query packet with id 10284 of size 56.
Freeing transaction 25325.
Processing incoming packet on transaction 10284 (rcode=NXDOMAIN).
Added NXDOMAIN cache entry for smbapptpgdb1.searchdomain3.com IN ANY 3600s
Transaction 10284 for <smbapptpgdb1.searchdomain3.com IN AAAA> on scope dns on eth0/* now complete with <rcode-failure> from network (unsigned).
Cache miss for smbapptpgdb1.searchdomain4.com IN AAAA
Transaction 41671 for <smbapptpgdb1.searchdomain4.com IN AAAA> scope dns on eth0/*.
Using feature level UDP+EDNS0 for transaction 41671.
Using DNS server 10.161.134.19 for transaction 41671.
Sending query packet with id 41671 of size 55.
Freeing transaction 10284.
Processing incoming packet on transaction 41671 (rcode=NXDOMAIN).
Added NXDOMAIN cache entry for smbapptpgdb1.searchdomain4.com IN ANY 3600s
Transaction 41671 for <smbapptpgdb1.searchdomain4.com IN AAAA> on scope dns on eth0/* now complete with <rcode-failure> from network (unsigned).
Positive cache hit for smbapptpgdb1.searchdomain5.com IN AAAA
Transaction 58104 for <smbapptpgdb1.searchdomain5.com IN AAAA> on scope dns on eth0/* now complete with <success> from cache (unsigned).
Following CNAME/DNAME smbapptpgdb1 → smbapptpgdb1.dhcp.searchdomain5.com.
NODATA cache hit for smbapptpgdb1.dhcp.searchdomain5.com IN AAAA
Transaction 21591 for <smbapptpgdb1.dhcp.searchdomain5.com IN AAAA> on scope dns on eth0/* now complete with <success> from cache (unsigned).
Freeing transaction 21591.
Sent message type=error sender=n/a destination=:1.336680 path=n/a interface=n/a member=n/a cookie=2043 reply_cookie=2 signature=s error-name=org.freedesktop.resolve1.NoSuchRR error-message='smbapptpgdb1.dhcp.searchdomain5.com' does not have any RR of the requested type
Sent message type=method_call sender=n/a destination=org.freedesktop.DBus path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=RemoveMatch cookie=2044 reply_cookie=0 signature=s error-name=n/a error-message=n/a
Freeing transaction 58104.
==215875== Invalid read of size 8
==215875==    at 0x13AF97: dns_query_candidate_go (resolved-dns-query.c:145)
==215875==    by 0x13B403: dns_query_candidate_notify (resolved-dns-query.c:277)
==215875==    by 0x141908: dns_transaction_complete (resolved-dns-transaction.c:351)
==215875==    by 0x1435B2: dns_transaction_process_dnssec (resolved-dns-transaction.c:839)
==215875==    by 0x1441C7: dns_transaction_process_reply (resolved-dns-transaction.c:1172)
==215875==    by 0x1445C3: on_dns_packet (resolved-dns-transaction.c:1224)
==215875==    by 0x4ACC1B8: source_dispatch (sd-event.c:3181)
==215875==    by 0x4ACDBEF: sd_event_dispatch (sd-event.c:3620)
==215875==    by 0x4ACE0C6: sd_event_run (sd-event.c:3678)
==215875==    by 0x4ACE29D: sd_event_loop (sd-event.c:3700)
==215875==    by 0x1168A5: run (resolved.c:92)
==215875==    by 0x116A06: main (resolved.c:99)
==215875==  Address 0x7f51de0 is 32 bytes inside a block of size 72 free'd
==215875==    at 0x483A9F5: free (vg_replace_malloc.c:538)
==215875==    by 0x13A37C: mfree.lto_priv.4 (alloc-util.h:48)
==215875==    by 0x13AAA3: dns_query_candidate_free (resolved-dns-query.c:66)
==215875==    by 0x13AB91: dns_query_candidate_unref (resolved-dns-query.c:69)
==215875==    by 0x13B585: dns_query_unref_candidates (resolved-dns-query.c:311)
==215875==    by 0x13D685: dns_query_cname_redirect (resolved-dns-query.c:933)
==215875==    by 0x13D98F: dns_query_process_cname (resolved-dns-query.c:978)
==215875==    by 0x123B93: bus_method_resolve_hostname_complete (resolved-bus.c:154)
==215875==    by 0x13C41D: dns_query_complete (resolved-dns-query.c:495)
==215875==    by 0x13D1EE: dns_query_accept (resolved-dns-query.c:825)
==215875==    by 0x13D31A: dns_query_ready (resolved-dns-query.c:858)
==215875==    by 0x13B422: dns_query_candidate_notify (resolved-dns-query.c:287)
==215875==  Block was alloc'd at
==215875==    at 0x483BAE9: calloc (vg_replace_malloc.c:760)
==215875==    by 0x13A637: dns_query_candidate_new (resolved-dns-query.c:24)
==215875==    by 0x13C571: dns_query_add_candidate (resolved-dns-query.c:515)
==215875==    by 0x13CD3A: dns_query_go (resolved-dns-query.c:689)
==215875==    by 0x1247FE: bus_method_resolve_hostname (resolved-bus.c:363)
==215875==    by 0x4A5CFA0: method_callbacks_run (bus-objects.c:415)
==215875==    by 0x4A601F1: object_find_and_run (bus-objects.c:1325)
==215875==    by 0x4A60870: bus_process_object (bus-objects.c:1445)
==215875==    by 0x4A7C6BE: process_message (sd-bus.c:2777)
==215875==    by 0x4A7C8B4: process_running.lto_priv.0 (sd-bus.c:2819)
==215875==    by 0x4A7D4DD: bus_process_internal (sd-bus.c:3039)
==215875==    by 0x4A7D5FC: sd_bus_process (sd-bus.c:3066)
```

### PR DESCRIPTION
**Pull request by keszybz**
In 0e0fd08fc832b8f42e567d722d388eba086da5ff I added reference counts to keep
track of the DnsQueryCandidate objects. Unfortunately, dns_query_unref_candidates()
was written as

     while (q->candidates)
           dns_query_candidate_unref(q->candidates);

i.e. it would keep dropping the reference count as many times as needed for it
to hit 0, making the patch less than fully effective.

dns_query_unref_candidates() is renamed to dns_query_detach_candidates() and
changed to drop exactly one reference from each of the linked candidates.

Example failure:
==463== Invalid read of size 8
==463==    at 0x419C93: dns_query_candidate_go (resolved-dns-query.c:159)
==463==    by 0x41A143: dns_query_candidate_notify (resolved-dns-query.c:304)
==463==    by 0x434BD6: dns_transaction_complete (resolved-dns-transaction.c:437)
==463==    by 0x436A0F: dns_transaction_process_dnssec (resolved-dns-transaction.c:976)
==463==    by 0x4378C1: dns_transaction_process_reply (resolved-dns-transaction.c:1387)
==463==    by 0x437CE9: on_dns_packet (resolved-dns-transaction.c:1444)
==463==    by 0x4B2DC9B: source_dispatch (sd-event.c:3512)
==463==    by 0x4B2FB1F: sd_event_dispatch (sd-event.c:4077)
==463==    by 0x4B2FFFA: sd_event_run (sd-event.c:4138)
==463==    by 0x4B301D6: sd_event_loop (sd-event.c:4159)
==463==    by 0x464A24: run (resolved.c:92)
==463==    by 0x464B3C: main (resolved.c:99)
==463==  Address 0x5f409d0 is 32 bytes inside a block of size 72 free'd
==463==    at 0x48410E4: free (vg_replace_malloc.c:755)
==463==    by 0x418EDF: mfree (alloc-util.h:48)
==463==    by 0x4197E8: dns_query_candidate_free (resolved-dns-query.c:67)
==463==    by 0x4198B7: dns_query_candidate_unref (resolved-dns-query.c:70)
==463==    by 0x41A2E3: dns_query_unref_candidates (resolved-dns-query.c:337)
==463==    by 0x41C5FE: dns_query_cname_redirect (resolved-dns-query.c:1028)
==463==    by 0x41CA04: dns_query_process_cname_one (resolved-dns-query.c:1128)
==463==    by 0x41CA80: dns_query_process_cname_many (resolved-dns-query.c:1157)
==463==    by 0x40C0BD: bus_method_resolve_hostname_complete (resolved-bus.c:198)
==463==    by 0x41B312: dns_query_complete (resolved-dns-query.c:562)
==463==    by 0x41C1AC: dns_query_accept (resolved-dns-query.c:922)
==463==    by 0x41C2C4: dns_query_ready (resolved-dns-query.c:955)
==463==    by 0x41A162: dns_query_candidate_notify (resolved-dns-query.c:314)
==463==    by 0x434BD6: dns_transaction_complete (resolved-dns-transaction.c:437)
==463==    by 0x438995: dns_transaction_prepare (resolved-dns-transaction.c:1728)
==463==    by 0x43921D: dns_transaction_go (resolved-dns-transaction.c:1928)
==463==    by 0x419C7C: dns_query_candidate_go (resolved-dns-query.c:163)
==463==    by 0x41A143: dns_query_candidate_notify (resolved-dns-query.c:304)
==463==    by 0x434BD6: dns_transaction_complete (resolved-dns-transaction.c:437)
==463==    by 0x436A0F: dns_transaction_process_dnssec (resolved-dns-transaction.c:976)
==463==    by 0x4378C1: dns_transaction_process_reply (resolved-dns-transaction.c:1387)
==463==    by 0x437CE9: on_dns_packet (resolved-dns-transaction.c:1444)
==463==    by 0x4B2DC9B: source_dispatch (sd-event.c:3512)
==463==    by 0x4B2FB1F: sd_event_dispatch (sd-event.c:4077)
==463==    by 0x4B2FFFA: sd_event_run (sd-event.c:4138)
==463==    by 0x4B301D6: sd_event_loop (sd-event.c:4159)
==463==    by 0x464A24: run (resolved.c:92)
==463==    by 0x464B3C: main (resolved.c:99)
==463==  Block was alloc'd at
==463==    at 0x483E86F: malloc (vg_replace_malloc.c:380)
==463==    by 0x418F81: malloc_multiply (alloc-util.h:96)
==463==    by 0x419378: dns_query_candidate_new (resolved-dns-query.c:23)
==463==    by 0x41B42C: dns_query_add_candidate (resolved-dns-query.c:582)
==463==    by 0x41BB7A: dns_query_go (resolved-dns-query.c:762)
==463==    by 0x40CE3A: bus_method_resolve_hostname (resolved-bus.c:464)
==463==    by 0x4A84B86: method_callbacks_run (bus-objects.c:414)
==463==    by 0x4A87961: object_find_and_run (bus-objects.c:1323)
==463==    by 0x4A87FEE: bus_process_object (bus-objects.c:1443)
==463==    by 0x4AA3434: process_message (sd-bus.c:2964)
==463==    by 0x4AA3623: process_running (sd-bus.c:3006)
==463==    by 0x4AA4110: bus_process_internal (sd-bus.c:3226)
==463==    by 0x4AA41EF: sd_bus_process (sd-bus.c:3253)
==463==    by 0x4AA5343: io_callback (sd-bus.c:3604)
==463==    by 0x4B2DC9B: source_dispatch (sd-event.c:3512)
==463==    by 0x4B2FB1F: sd_event_dispatch (sd-event.c:4077)
==463==    by 0x4B2FFFA: sd_event_run (sd-event.c:4138)
==463==    by 0x4B301D6: sd_event_loop (sd-event.c:4159)
==463==    by 0x464A24: run (resolved.c:92)
==463==    by 0x464B3C: main (resolved.c:99)

Fixes #19376.

### CODE REVIEW

File: src/resolve/resolved-dns-query.c [hunk 1]
```diff
@@ -42,28 +42,43 @@ static void dns_query_candidate_stop(DnsQueryCandidate *c) {
...
+        /* Detach all the DnsTransactions attached to this query */
+
         while ((t = set_steal_first(c->transactions))) {
                 set_remove(t->notify_query_candidates, c);
                 set_remove(t->notify_query_candidates_done, c);
                 dns_transaction_gc(t);
         }
 }
 
+static DnsQueryCandidate* dns_query_candidate_detach(DnsQueryCandidate *c) {
```
Comment (poettering) [inline src/resolve/resolved-dns-query.c]: hmm, i think there's prior art in our codebase to call funcitons like this "xyz_unlink()", but doesn#t matter

File: src/resolve/resolved-dns-query.c [hunk 2]
```diff
@@ -330,11 +350,13 @@ static void dns_query_stop(DnsQuery *q) {
                 dns_query_candidate_stop(c);
 }
 
-static void dns_query_unref_candidates(DnsQuery *q) {
+static void dns_query_detach_candidates(DnsQuery *q) {
```
Comment (poettering) [inline src/resolve/resolved-dns-query.c]: as above, maybe detach → unlink, but doesn't matter

**Review by poettering — APPROVED** [review]
lgtm. two suggestions

**Comment by keszybz** [comment]
Updated with detach → unlink rename.

**Pull request merged**